### PR TITLE
add function for cancelling all active block orders for a given market

### PR DESCRIPTION
**Pull request by martineehrlich**
## Description
When a user releases their channels, we should cancel all their orders on that market because the user will no longer have the funds to fulfill those orders.

## Todos
- [x] Tests
- [x] Documentation
- [x] Link to Trello

### CODE REVIEW

**Comment by martineehrlich** [comment]
[Cancel orders when releasing channels](https://trello.com/c/ieLGxqrG/1131-cancel-orders-when-releasing-channels)

**Comment by martineehrlich** [comment]
I still need to manually test and I am thinking that the naming for successfulCancellations and unsuccessfulCancellations could change.

Also I am aware that getBlockOrders finds the blockorders, and then I call cancelBlockOrder with the blockOrder.id, which gets the record from the database again. It didn't really make sense to refactor the cancelBlockOrder code which requires an id for the sake of not double fetching on a call that will never be huge anyways. 

Also, for releaseChannels, I wasn't sure what we wanted the behavior to be. Currently I am just logging because I wasn't sure if it made sense to pass error messages about order cancellation to the user on channel closure.

File: broker-daemon/block-order-worker/index.js [hunk 1]
```diff
@@ -358,6 +358,31 @@ class BlockOrderWorker extends EventEmitter {
...
+    this.logger.info('Cancelling all active orders for market', { market })
+
+    const blockOrders = await this.getBlockOrders(market)
+    const activeBlockOrders = blockOrders.filter(blockOrder => blockOrder.isActive)
+    const successfulCancellations = []
+    const unsuccessfulCancellations = []
+
+    for (let blockOrder of activeBlockOrders) {
+      try {
+        await this.cancelBlockOrder(blockOrder.id)
```
Comment (treygriffith) [inline broker-daemon/block-order-worker/index.js]: having this `await` in the loop will cause each of these to cancel in parallel, which seems unnecessarily slow.

File: broker-daemon/block-order-worker/index.js [hunk 2]
```diff
@@ -358,6 +358,31 @@ class BlockOrderWorker extends EventEmitter {
...
+        successfulCancellations.push(blockOrder.id)
+      } catch (e) {
+        unsuccessfulCancellations.push(blockOrder.id)
```
Comment (treygriffith) [inline broker-daemon/block-order-worker/index.js]: I think we need to capture the error here somehow, either by returning it or logging it

File: broker-daemon/block-order-worker/index.spec.js [hunk 3]
```diff
@@ -1211,6 +1211,73 @@ describe('BlockOrderWorker', () => {
     })
   })
 
+  describe.only('#cancelActiveOrders', () => {
```
Comment (treygriffith) [inline broker-daemon/block-order-worker/index.spec.js]: make sure to lose this only before merging

File: broker-daemon/broker-rpc/wallet-service/release-channels.js [hunk 4]
```diff
@@ -84,6 +84,12 @@ async function releaseChannels ({ params, logger, engines, orderbooks }, { Relea
     closeChannels(counterEngine, counterSymbol, force, logger)
   ])
 
+  if (base.status === RELEASE_STATE.RELEASED && counter.status === RELEASE_STATE.RELEASED) {
+    const { successfulCancellations, unsuccessfulCancellations } = await blockOrderWorker.cancelActiveOrders(market)
```
Comment (treygriffith) [inline broker-daemon/broker-rpc/wallet-service/release-channels.js]: Probably want this to happen before, or in tandem with the channels closing. One of the channels failing to close resulting in the orders not being cancelled is a bit of weird behavior I think.

File: broker-daemon/block-order-worker/index.js [hunk 5]
```diff
@@ -358,6 +358,34 @@ class BlockOrderWorker extends EventEmitter {
...
+    const cancelledOrders = []
+    const failedToCancelOrders = []
+
+    await Promise.all(
+      activeBlockOrders.map(async (blockOrder) => {
+        try {
+          await this.cancelBlockOrder(blockOrder.id)
+          cancelledOrders.push(blockOrder.id)
+        } catch (e) {
+          this.logger.error('Failed to cancel order', e)
```
Comment (treygriffith) [inline broker-daemon/block-order-worker/index.js]: probably need the blockorder id in the logger. Also - if you pass the error as the second parameter to the logger it doesn't work well, it's better like:

```
this.logger.error('Failed to cancel block order', { blockOrderId: blockOrder.id, error: e })
```

File: broker-daemon/broker-rpc/wallet-service/release-channels.js [hunk 6]
```diff
@@ -67,6 +67,10 @@ async function releaseChannels ({ params, logger, engines, orderbooks }, { Relea
     throw new PublicError(`${market} is not being tracked as a market.`)
   }
 
+  const { cancelledOrders, failedToCancelOrders } = await blockOrderWorker.cancelActiveOrders(market)
+  logger.info('Successfully cancelled orders', { orders: cancelledOrders })
+  logger.info('Failed to cancel orders', { orders: failedToCancelOrders })
```
Comment (treygriffith) [inline broker-daemon/broker-rpc/wallet-service/release-channels.js]: weird to have this log if no block orders failed to cancel

File: broker-daemon/block-order-worker/index.js [hunk 7]
```diff
@@ -358,6 +358,34 @@ class BlockOrderWorker extends EventEmitter {
     return blockOrder
   }
 
+  /**
+   * Cancel all active orders for a given market
+   * @param  {string} market to cancel orders on
+   * @return {object} cancelledOrders and failedToCancelOrders as arrays of blockorderids
```
Comment (dannypaz) [inline broker-daemon/block-order-worker/index.js]: For JSDoc, the representation of your return value should look like:

```
@returns {object} result
@returns {Array<string>} result.cancelledOrders ...
etc.. etc..
```

File: broker-daemon/block-order-worker/index.js [hunk 8]
```diff
@@ -358,6 +358,36 @@ class BlockOrderWorker extends EventEmitter {
...
+          await this.cancelBlockOrder(blockOrder.id)
+          cancelledOrders.push(blockOrder.id)
+        } catch (e) {
+          this.logger.error('Failed to cancel block order', { blockOrderId: blockOrder.id, error: e })
+          failedToCancelOrders.push(blockOrder.id)
+        }
+      })
+    )
+
+    return { cancelledOrders, failedToCancelOrders }
```
Comment (treygriffith) [inline broker-daemon/block-order-worker/index.js]: [nit] might be nice to log the number of cancelled orders (and failed)

**Pull request merged**